src/tools/pbs_upgrade_job.c

```diff
@@ -234,102 +311,284 @@ check_job_file(int fd)
 	return ret_version;
 }
 
+/**
+ * @brief	Upgrade pre 18.x jobfix structure to 19
+ *
+ * @param[in]	old_jobfix_pre19 - pre-19 jobfix struct
+ *
+ * @return	jobfix_19_20
+ * @retval	v19 converted jobfix
+ */
+jobfix_19_20
+convert_pre18jf_to_19(jobfix_PRE19 old_jobfix_pre19)
+{
+	jobfix_19_20 jf_19_20;
+
+	/* Copy the data to the new jobfix structure */
+	memset(&jf_19_20, 0, sizeof(jf_19_20));
+	memcpy(&jf_19_20, &old_jobfix_pre19, sizeof(jf_19_20));
+	snprintf(jf_19_20.ji_jobid, sizeof(jf_19_20.ji_jobid),
+			"%s", old_jobfix_pre19.ji_jobid);
+	snprintf(jf_19_20.ji_fileprefix, sizeof(jf_19_20.ji_fileprefix),
+			"%s", old_jobfix_pre19.ji_fileprefix);
+	snprintf(jf_19_20.ji_queue, sizeof(jf_19_20.ji_queue),
+			"%s", old_jobfix_pre19.ji_queue);
+	snprintf(jf_19_20.ji_destin, sizeof(jf_19_20.ji_destin),
+			"%s", old_jobfix_pre19.ji_destin);
+	jf_19_20.ji_un_type = old_jobfix_pre19.ji_un_type;
+	memcpy(&jf_19_20.ji_un, &old_jobfix_pre19.ji_un, sizeof(jf_19_20.ji_un));
+
+	return jf_19_20;
+}
+
+/**
+ * @brief	Upgrade 19-20 jobfix structure to latest
+ *
+ * @param[in]	old_jf - 19-20 jobfix struct
+ *
+ * @return	jobfix_19_20
+ * @retval	v19 converted jobfix
+ */
+struct jobfix
+convert_19jf_to_21(jobfix_19_20 old_jf)
+{
+	struct jobfix jf;
+
+	memset(&jf, 0, sizeof(jf));
+	jf.ji_jsversion = JSVERSION;
+	jf.ji_svrflags = old_jf.ji_svrflags;
+	jf.ji_stime = old_jf.ji_stime;
+	snprintf(jf.ji_jobid, sizeof(jf.ji_jobid), "%s", old_jf.ji_jobid);
+	snprintf(jf.ji_fileprefix, sizeof(jf.ji_fileprefix), "%s", old_jf.ji_fileprefix);
+	snprintf(jf.ji_queue, sizeof(jf.ji_queue), "%s", old_jf.ji_queue);
+	snprintf(jf.ji_destin, sizeof(jf.ji_destin), "%s", old_jf.ji_destin);
+	jf.ji_un_type = old_jf.ji_un_type;
+	memcpy(&jf.ji_un, &old_jf.ji_un, sizeof(jf.ji_un));
+
+	return jf;
+}
+
+union jobextend
+convert_19ext_to_21(union jobextend_19_20 old_extend)
+{
+	union jobextend je;
+
+	memset(&je, 0, sizeof(je));
+	snprintf(je.fill, sizeof(je.fill), "%s", old_extend.fill);
+	snprintf(je.ji_ext.ji_jid, sizeof(je.ji_ext.ji_jid), "%s", old_extend.ji_ext.ji_4jid);
```

Comment:
jobextend_19_20 has conditional compilation block for __sgi. Don't you need it here too while copying jobid? If we don't support __sgi is the newer build then happens when the upgrade is made on __sgi compatible system.
If __sgi is completely not supported then I wouldn't keep it in jobextend_19_20 structure as well.